From e9d4afd0f41782133fcdb72648d8510c767b1448 Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Sat, 03 Oct 2020 13:57:57 +0300
Subject: [PATCH] [SW-20984] goya2: add support for extended btl fit

Goya2 onwards btl image is extended to have 6 parts.
This patch adds new uboot_extended.its file which
adds these extended sections to boot fit file. fw
loader configuration is extended to include these
images in flash programming.

Change-Id: I5257dd30abab226447fda1da6aa366a84e6ea6c1
---

diff --git a/Config.in b/Config.in
index 8af0698..aac3232 100644
--- a/Config.in
+++ b/Config.in
@@ -6,6 +6,7 @@
 source "$BR2_EXTERNAL_HABANALABS_PATH/package/eeprom-util/Config.in"
 source "$BR2_EXTERNAL_HABANALABS_PATH/package/arcboot/Config.in"
 source "$BR2_EXTERNAL_HABANALABS_PATH/package/zephyr/Config.in"
+source "$BR2_EXTERNAL_HABANALABS_PATH/package/post-image/Config.in"
 
 source "$BR2_EXTERNAL_HABANALABS_PATH/package/openbmc/Config.in"
 
diff --git a/board/habanalabs/common/post-image-fit.sh b/board/habanalabs/common/post-image-fit.sh
index d944bc5..99c165e 100755
--- a/board/habanalabs/common/post-image-fit.sh
+++ b/board/habanalabs/common/post-image-fit.sh
@@ -36,7 +36,7 @@
 COMMON_BASE=${BR2_EXTERNAL_HABANALABS_PATH}/board/habanalabs/common
 UBOOT_ITS_FILE=uboot-${HL_DEV}_${HL_TARGET}.its
 UBOOT_ITB_FILE=uboot-fit.itb
-UBOOT_ITS_TEMPLATE=${COMMON_BASE}/uboot.its
+UBOOT_ITS_TEMPLATE=`sed -e 's/^"//' -e 's/"$//' <<< ${BR2_POST_IMAGE_UBOOT_ITS_TEMPLATE}`
 SPI_ITS_FILE=spi-${HL_DEV}_${HL_TARGET}.its
 SPI_ITB_FILE=spi-fpga-fit.itb
 SPI_FPGA_ITS_TEMPLATE=${COMMON_BASE}/spi-fpga-legacy.its
diff --git a/board/habanalabs/common/uboot_extended.its b/board/habanalabs/common/uboot_extended.its
new file mode 100644
index 0000000..e9ae8b5
--- /dev/null
+++ b/board/habanalabs/common/uboot_extended.its
@@ -0,0 +1,176 @@
+/*
+ * Simple U-Boot uImage source file containing a
+ * single U-Boot blob
+ */
+
+/dts-v1/;
+
+/ {
+	hl-version = "@HL_VERSION@";
+	hl-uboot-version = "@HL_UBOOT_VERSION@";
+	hl-preboot-version = "@HL_PREBOOT_VERSION@";
+	hl-btl-version = "@HL_BTL_VERSION@";
+	hl-device = "@HL_DEV@";
+	hl-sec-key-idx = <@HL_KEY_INDEX@>;
+	hl-sec-version = <@HL_SEC_FW_VERSION@>;
+	description = "FIT image with U-Boot, preboot and proper BTL";
+	#address-cells = <2>;
+
+	images {
+		btl-s1 {
+			description = "Habana BTL primary";
+			data = /incbin/("s1.bin.le");
+			compression = "none";
+			offset = <0x00000000>;
+			hash {
+				algo = "sha256";
+			};
+		};
+
+		btl-s2 {
+			description = "Habana BTL secondary";
+			data = /incbin/("s2.bin.le");
+			compression = "none";
+			offset = <0x00010000>;
+			hash {
+				algo = "sha256";
+			};
+		};
+
+		btl-s3 {
+			description = "Habana BTL prstn 1";
+			data = /incbin/("s3.bin.le");
+			compression = "none";
+			offset = <0x00020000>;
+			hash {
+				algo = "sha256";
+			};
+		};
+
+		btl-s4 {
+			description = "Habana BTL prstn 2";
+			data = /incbin/("s4.bin.le");
+			compression = "none";
+			offset = <0x00030000>;
+			hash {
+				algo = "sha256";
+			};
+		};
+
+		btl-s5 {
+			description = "Habana BTL pcie 1";
+			data = /incbin/("s5.bin.le");
+			compression = "none";
+			offset = <0x00040000>;
+			hash {
+				algo = "sha256";
+			};
+		};
+
+		btl-s6 {
+			description = "Habana BTL pcie 2";
+			data = /incbin/("s6.bin.le");
+			compression = "none";
+			offset = <0x00050000>;
+			hash {
+				algo = "sha256";
+			};
+		};
+
+		img-ptr {
+			description = "Habana image pointers";
+			data = /incbin/("@HL_DEV@-images-pointers.bin.le");
+			compression = "none";
+			offset = <0x00080000>;
+			hash {
+				algo = "sha256";
+			};
+		};
+
+		uboot-env {
+			description = "Habana U-Boot environment";
+			data = /incbin/("@HL_DEV@-uboot-env.bin.le");
+			compression = "none";
+			offset = <0x000A0000>;
+			hash {
+				algo = "sha256";
+			};
+		};
+
+		preboot {
+			description = "Habana Preboot";
+			data = /incbin/("@HL_DEV@-preboot.bin.le");
+			type = "standalone";
+			arch = "arm64";
+			compression = "none";
+			offset = <0x000B0000>, <0x000C0000>;
+			loadaddr = <0x7F 0xF0004000>;
+			hash {
+				algo = "sha256";
+			};
+		};
+
+		uboot {
+			description = "U-Boot (64-bit)";
+			data = /incbin/("@HL_DEV@-u-boot.bin.le");
+			type = "standalone";
+			arch = "arm64";
+			compression = "none";
+			loadaddr = <0x7F 0xF0100000>;
+			hash {
+				algo = "sha256";
+			};
+		};
+
+		pub-key {
+			description = "Public key in DTB format";
+			data = /incbin/("@KEY_DTB@");
+			type = "standalone";
+			compression = "none";
+			hash {
+				algo = "sha256";
+			};
+		};
+	};
+
+	configurations {
+		default = "fw-loader-conf";
+
+		fw-loader-conf {
+			load-type = <@FIT_LOAD_TYPE_FLASH@>;
+			description = "Habanalabs device boot loader";
+			loadables = "btl-s1", "btl-s2", "btl-s3", "btl-s4",
+				  "btl-s5" , "btl-s6", "img-ptr", "uboot-env",
+				  "preboot";
+			btl-s1 = "btl-s1";
+			btl-s2 = "btl-s2";
+			btl-s3 = "btl-s3";
+			btl-s4 = "btl-s4";
+			btl-s5 = "btl-s5";
+			btl-s6 = "btl-s6";
+			img-ptr = "img-ptr";
+			uboot-env = "uboot-env";
+			preboot = "preboot";
+			pub-key = "pub-key";
+			signature {
+				algo = "sha256,rsa3072";
+				key-name-hint = "hl_rsa3072";
+				sign-images = "btl-s1", "btl-s2", "btl-s3",
+					"btl-s4", "btl-s5", "btl-s6",
+					"img-ptr", "uboot-env", "preboot";
+			};
+		};
+
+		preboot-conf {
+			description = "Habanalabs device preboot";
+			loadables = "uboot";
+			uboot = "uboot";
+			pub-key = "pub-key";
+			signature {
+				sign-images = "uboot";
+				algo = "sha256,rsa3072";
+				key-name-hint = "hl_rsa3072";
+			};
+		};
+	};
+};
diff --git a/configs/goya2_defconfig b/configs/goya2_defconfig
index 423cc2d..3cc9074 100644
--- a/configs/goya2_defconfig
+++ b/configs/goya2_defconfig
@@ -9,6 +9,7 @@
 BR2_TARGET_GENERIC_ROOT_PASSWD="root"
 BR2_SYSTEM_BIN_SH_BASH=y
 BR2_SYSTEM_DHCP="eth0"
+BR2_POST_IMAGE_UBOOT_ITS_TEMPLATE="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/common/uboot_extended.its"
 BR2_ROOTFS_OVERLAY="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/common/rootfs_overlay"
 BR2_ROOTFS_POST_BUILD_SCRIPT="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/common/post-build.sh"
 BR2_ROOTFS_POST_IMAGE_SCRIPT="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/common/post-image-spi.sh $(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/common/post-image-fit.sh"
diff --git a/package/post-image/Config.in b/package/post-image/Config.in
new file mode 100644
index 0000000..830f577
--- /dev/null
+++ b/package/post-image/Config.in
@@ -0,0 +1,13 @@
+config BR2_PACKAGE_POST_IMAGE
+	bool
+	default y
+	help
+	  virtual package to override post image
+	  processing configurations.
+
+config BR2_POST_IMAGE_UBOOT_ITS_TEMPLATE
+	string "Path to template of uboot its file"
+	default "$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/common/uboot.its"
+	help
+	  Path to its template file used in creation
+	  of project specific boot fit.
diff --git a/package/post-image/post-image.mk b/package/post-image/post-image.mk
new file mode 100644
index 0000000..9789ec1
--- /dev/null
+++ b/package/post-image/post-image.mk
@@ -0,0 +1 @@
+export BR2_POST_IMAGE_UBOOT_ITS_TEMPLATE
